File: 0124-binary-tree-maximum-path-sum/0124-binary-tree-maximum-path-sum.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int maxPathSum(TreeNode* root) {
        int maxValue=INT_MIN;
        maxPathDown(root,maxValue);
        return maxValue;
    }
private:
    int maxPathDown(TreeNode *root,int& maxValue){
        if(root == NULL) return 0;
        int leftSum=max(0,maxPathDown(root->left,maxValue));
        int rightSum= max(0, maxPathDown(root->right,maxValue));
        maxValue=max(maxValue, (leftSum+rightSum)+root->val);
        return root->val+max(leftSum,rightSum);
    }
};

File: 0124-binary-tree-maximum-path-sum/0124-binary-tree-maximum-path-sum.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public int maxPathSum(TreeNode root) {
        int[] maxValue= new int[1];
        maxValue[0] = Integer.MIN_VALUE;
        maxPathDown(root, maxValue); 
        return maxValue[0];
    }
    private int maxPathDown(TreeNode root, int maxValue[]){
        if(root == null) return 0;
        int leftSum=Math.max(0,maxPathDown(root.left,maxValue));
        int rightSum= Math.max(0, maxPathDown(root.right,maxValue));
        maxValue[0]=Math.max(maxValue[0], (leftSum+rightSum)+root.val);
        return root.val+Math.max(leftSum,rightSum);
    }
    
}